### PR TITLE
THE DOOR OPENING...ENING... (IDs can send access requests to door remotes, misc door remote additions and changes)

### PR DESCRIPTION
**Pull request by Metekillot**
## About The Pull Request

Door remotes are closer to clunky tchotchkes meant more to serve as a bit of impractical bling for Heads of Staff rather than something they'd make practical use of. Or rather, that seems what they're meant for if their actual use is any indication.

This pull request adds a system for ID cards to send an access request to door remotes. The holder of the remote will get a little ding, and the person with the ID will be told to sit there and wait for the Head of Staff's pleasure like the prole they are.

Ideally, we might take door opening away from AIs to foist a pain in the ass responsibility on Heads of Staff instead, in a way that is meant to encourage them to do ID access updates for people they want coming and going from their department frequently who don't already have access. Paramedics being Assistant+Access better watch your backs.

Also, door remotes will be able to shock doors if you emag them, because lol, and down the line I'd intended to have door remotes be upgradeable but that's outside the scope of this particular PR.

Heads of Staff can respond to an access request with:

- Approve (open the door)
- Set door to EA
- Reject (temporarily blocks more requests from that ID)
- Bolt door (when you want to be a passive aggressive diva)
- Block ID (you will no longer receive access requests from that ID, and they'll be told that you blocked them if they try again)
- Clear requests (oh, it must have gotten lost in my uh... paperwork?)

They can set any access request to be replied to automatically with a chosen option from these. That is as stupid an idea as it sounds but the Clown stealing the Captain's all-access remote and setting it to auto-bolt any requests that get routed up to it sounds really funny.

Oh, I should also mention that with this, I had intended (depending on how violent of a knee-jerk a change like this elicits)
- For the Captain to spawn in with an all-access remote
- Or, if the Captain is round-start, let them choose to have the spare ID instead be this all-access remote

Either one encourages people to play Captain, and to play Captain roundstart if the second one. It would significantly cut down on all-access acting Captain research director heretic, or the 500th acting captain who has decided that SYNDICATE CAPTAIN is fucking hilarious, however that's more a community issue than a mechanical one. If there's no roundstart Captain and no Captain joins the round, this remote does not spawn.

The holder of a remote can also just turn request listening for the remote off if they feel lazy and still want to make AI DOOR.

The holder of a door remote can also choose to be notified if the AI opens a door they have access to.

The Captain will only receive access requests to their Very Special Head remote if a given Head of staff specifically elects to route cleared requests up; this will not happen automatically without the involvement of a department's head of staff... because that's definitely the only person who would have their remote.

### Door remote has a radial menu now!

Roundstart heads of staff start with their remote in their backpack.
## Why It's Good For The Game

Giving Heads of Staff a little more petty power in their chosen departmental fiefdom always seems like a good idea to me, personally.
Every day, a door remote gets forgotten in a locker. For just one merge a PR, you can house an abandoned door remote in a Head of Staff's pocket.
Making it practical for people to manage access to their department more easily will encourage Heads of Staff to coordinate their departments, rather than roll in to Research Director so they can be be Scientist with roundstart gamer loot
Captain losing the all-access door remote seems like it would be a catastrophic shitshow and I'm all for that
Non-Asimov AIs pretending not to hear your AI DOOR isn't quite as crippling, and if they still want to give you a hard time they can just bolt the door and depower it.
## Changelog
:cl: Bisar
add: Nanotrasen door remotes have received a computing update. In an effort to statistically lower the frequency of Malfunctioning AI incidents due to stress, ID cards can now attempt to route an access request up to the remote. The Head of Staff this remote has been assigned to may handle the access request as they see fit. Damaged remotes can be bluespaced to your local Nanotrasen Door Remote Repair Outpost and you will receive a new one in 3-5 business months. Do not attempt to use a damaged door remote.
add: The Captain spawns with an all-access door remote in their backpack if they join the shift late OR The Captain can choose to have the spare ID be replaced by an all-access door remote, but only if they join roundstart.
balance: Heads of staff have their remote in their backpack instead if they join roundstart.
qol: Door remote has a radial menu to change modes instead of cycling; thumbs rejoice.
/:cl:

### CODE REVIEW

File: code/__DEFINES/dcs/signals/signals_key.dm [hunk 1]
```diff
@@ -1,2 +1,8 @@
 ///Called when a keycard is sending a department's access. (obj/machinery/keycard_auth/source, list/region_access)
 #define COMSIG_ON_DEPARTMENT_ACCESS "on_department_access"
+
+///Called when someone sends a signal to any listening door remotes. (obj/item/card/id/ID_requesting, obj/machinery/door/airlock/requested_door)
+#define COMSIG_DOOR_REMOTE_ACCESS_REQUEST "door_remote_access_request"
+	#define COMPONENT_REQUEST_LIMIT_REACHED 1
+	#define COMPONENT_NOT_ON_THE_LIST_PAL 2
+	#define COMPONENT_REQUEST_RECEIVED 3
```
Comment (MrMelbert) [inline code/__DEFINES/dcs/signals/signals_key.dm:8]: Signals should only return bitflags, so these should be `1<<0` `1<<1` `1<<2`. 
May not functionally change it much ultimately.

File: code/controllers/subsystem/id_access.dm [hunk 2]
```diff
@@ -35,6 +37,8 @@ SUBSYSTEM_DEF(id_access)
 
 	/// The roundstart generated code for the spare ID safe. This is given to the Captain on shift start. If there's no Captain, it's given to the HoP. If there's no HoP
 	var/spare_id_safe_code = ""
+	/// Associated list of regions; entries are null until getting LAZYADD'd a ref to a given remote once it gets registered
+	var/remotes_listening_by_region = list()
```
Comment (MrMelbert) [inline code/controllers/subsystem/id_access.dm:41]:
```suggestion
	var/list/remotes_listening_by_region = list()
```

File: code/controllers/subsystem/id_access.dm [hunk 3]
```diff
@@ -506,3 +520,28 @@ SUBSYSTEM_DEF(id_access)
...
+ * * door_requested - The door that the ID card is requesting be opened.
+ */
+/datum/controller/subsystem/id_access/proc/route_request_to_door_remote(obj/item/card/id/ID_requesting, obj/machinery/door/airlock/door_requested)
+	. = FALSE
+	for(var/region in SSid_access.station_regions)
+		if(door_requested.check_access_list(SSid_access.accesses_by_region[region]))
+			for(var/obj/item/door_remote/remote in SSid_access.remotes_listening_by_region[region])
+				. = SEND_SIGNAL(remote, COMSIG_DOOR_REMOTE_ACCESS_REQUEST, ID_requesting, door_requested)
+	if(!.)
+		ID_requesting.visible_message("A scroll of text rolls across the front of [ID_requesting]: ACCESS REQUEST ROUTING FAILED, CONSULT ARTIFICIAL INTELLIGENCE FOR ASSISTANCE.", vision_distance = 1)
```
Comment (MrMelbert) [inline code/controllers/subsystem/id_access.dm:545]: Give this a `span`?

File: code/game/objects/items/cards_ids.dm [hunk 4]
```diff
@@ -454,11 +454,29 @@
...
+		. = validate_access_request_attempt(user, interacting_with)
+		return (. ? ITEM_INTERACT_SUCCESS : NONE)
+	if(!isfloorturf(interacting_with))
 		return NONE
 	try_project_paystand(user, interacting_with)
 	return ITEM_INTERACT_SUCCESS
 
+/obj/item/card/id/proc/validate_access_request_attempt(mob/living/user, obj/machinery/door/airlock/to_request)
+	. = FALSE
+	if(to_request.allowed(user) && !to_request.locked) // if you can use the door don't waste time asking for access to it
```
Comment (MrMelbert) [inline code/game/objects/items/cards_ids.dm:469]: This use of `.` is very superfluous, just `return FALSE` `return TRUE`

File: code/game/objects/items/control_wand.dm [hunk 5]
```diff
@@ -17,21 +18,107 @@
...
 	var/region_access = REGION_GENERAL
 	var/list/access_list
+	/// The name that gets sent back to IDs that send access requests to this remote. Defaults to department.
+	var/response_name = null
+	var/listening = FALSE
+	/// A list of paired items, the first being the ID card requesting access, the second being the door that access is requested for.
+	/// They'll only be able to request one door per ID, both so we're not cramming this full of lists that need to be GC'd and to make
+	/// remote requests kind of a pain in the ass and a situation where they should request adding the access to their ID from the relevant
+	/// head of staff.
+	var/open_requests = list()
```
Comment (MrMelbert) [inline code/game/objects/items/control_wand.dm:28]:
```suggestion
	var/list/open_requests = list()
```
Also maybe type this depending on contents I can't tell what's actually in this
```suggestion
	var/list/obj/item/open_requests = list()
```

File: code/game/objects/items/control_wand.dm [hunk 6]
```diff
@@ -17,21 +18,107 @@
...
+	else
+		toggle_listen(new_hand_touches_the_beacon)
+	UnregisterSignal(src, COMSIG_ITEM_PICKUP)
+
+/obj/item/door_remote/proc/toggle_listen(mob/user, setting_toggle = FALSE)
+	if(!setting_toggle)
+		if(!listening)
+			listening = !listening
+			RegisterSignal(src, COMSIG_DOOR_REMOTE_ACCESS_REQUEST, PROC_REF(receive_access_request))
+			log_admin("[src] has started listening for door access requests due to being picked up.")
```
Comment (MrMelbert) [inline code/game/objects/items/control_wand.dm:63]: Seems like a weird admin log

File: code/game/objects/items/control_wand.dm [hunk 7]
```diff
@@ -40,7 +127,77 @@
...
+			if(smile_youre_on_camera.is_operational && (get_dist(smile_youre_on_camera, requested_door) <= smile_youre_on_camera.view_range))
+				nearest_camera = smile_youre_on_camera
+				break
+		parsed_requests += "[request_item.registered_name] -> [open_requests[request_item]] ([nearest_camera ? nearest_camera.c_tag : get_area(requested_door)])"
+	parsed_requests += list(
+		"\[OPERATION\] Clear all requests",
+		"\[OPERATION\] [listening ? "Stop" : "Start"] listening for requests",
+		"\[OPERATION\] !CAUTION!_Set automatic request response_!CAUTION!)")
+	var/list/choices = list()
+	choices = tgui_input_checkboxes(user = user,
```
Comment (MrMelbert) [inline code/game/objects/items/control_wand.dm:157]:
```suggestion
	var/list/choices = tgui_input_checkboxes(user = user,
```

File: code/game/objects/items/control_wand.dm [hunk 8]
```diff
@@ -40,7 +127,77 @@
...
+	message = "Please choose either a single operation, or a set of door access requests to perform\
+	 a batch operation on. Nanotrasen Incorporated not liable for any criminal activity or loss of\
+	 profit and/or life resulting from door requests.",
+	   title = "DOOR REQUEST HANDLER AND CONFIGURATION", items = parsed_requests)
+	var/number_of_choices = length(choices)
+	if(!number_of_choices)
+		return
+	else
```
Comment (MrMelbert) [inline code/game/objects/items/control_wand.dm:165]: We're already early returning this, no need for `else`

File: code/game/objects/items/control_wand.dm [hunk 9]
```diff
@@ -113,40 +270,51 @@
 	department = "omni"
 	region_access = REGION_ALL_STATION
 
-/obj/item/door_remote/captain
+/obj/item/door_remote/command
 	name = "command door remote"
 	department = "command"
+	response_name = span_comradio("CAPTAIN")
```
Comment (MrMelbert) [inline code/game/objects/items/control_wand.dm:276]: For bonus points grab these titles from the job datums:
`response_name = span_comradio(/datum/job/captain::title)` or whatever the var is called

File: code/modules/jobs/job_types/captain.dm [hunk 10]
```diff
@@ -106,7 +107,16 @@
...
 	. = ..()
-	if(visualsOnly || !special_charter)
+	if(visualsOnly)
+		return
+
+	var/obj/item/door_remote/omni/cap_remote = locate() in equipped.get_item_by_slot(ITEM_SLOT_BACK)
+	if(isnull(cap_remote))
+		// failed to give out the remote, plop on the ground
+		cap_remote = new(get_turf(equipped))
+	cap_remote.name = "Captain's door remote"
```
Comment (MrMelbert) [inline code/modules/jobs/job_types/captain.dm:117]: Why is this necessary?

Comment (Metekillot) [reply code/modules/jobs/job_types/captain.dm:117]: The balance change or this bit where it gets put in the backpack?

**Comment by Admiral-Gigaton** [comment]
taking door openning from AIs is like 50% of AI content, but also it annoys AI players to open doors.

I know its not the direct purpose of this PR, but god no we should never remove "AI DOOR" because that is like... core human to "human" interactions on the station (beating the shit head who broke in with AI or askign AI to open)

**Comment by carlarctg** [comment]
> taking door openning from AIs is like 50% of AI content, but also it annoys AI players to open doors.
> 
> I know its not the direct purpose of this PR, but god no we should never remove "AI DOOR" because that is like... core human to "human" interactions on the station (beating the shit head who broke in with AI or askign AI to open)

if door opening is the only thing ais meaningfully contribute to a round in a positive way then maybe they shouldn't exist 


10 oudda ten pr, absolutely amazing, love the choose spare or remote idea.

**Comment by carlarctg** [comment]
notes:


add some examine help on ids and remotes and airlocks for visibility 

think long and hard on how to make this feature visible to players

i love how this pr mixes wol with amazing new features that encourage good head support and even also goofy features that are more prank than practical

**Comment by GuillaumePrata** [comment]
> The holder of a door remote can also choose to be notified if the AI opens a door they have access to.

I think this is a bad idea, people using silicons to sneak around departments adds to the round and heads of staff acting like security is just lame.

If command thinks someone is abusing silicons to get illicit access, they can change the laws, get a detective to look for evidence and other options. Just have an instant "There is someone using silicons to break into this room" just encourages command to be valid hunters.

**Comment by Metekillot** [comment]
> > taking door openning from AIs is like 50% of AI content, but also it annoys AI players to open doors.
> > 
> > I know its not the direct purpose of this PR, but god no we should never remove "AI DOOR" because that is like... core human to "human" interactions on the station (beating the shit head who broke in with AI or askign AI to open)
> 
> if door opening is the only thing ais meaningfully contribute to a round in a positive way then maybe they shouldn't exist 

Or we could make new things for them to do instead, or let them be more involved that things that need doing now.

**Comment by EricZilla** [comment]
> departmental door boards that shows what doors are what status in a later PR.

You open up a TGUI menu with the remote but its just the FNAF 3 vent menu

File: code/modules/jobs/job_types/captain.dm (hunk 10 above)
Comment (MrMelbert) [reply code/modules/jobs/job_types/captain.dm:117]: This code to put it in the backpack and rename it. You already use the `backpack_contents` list, so why is this here?

**Comment by xzero314** [comment]
-Ideally, we might take door opening away from AIs - Why? Its one of the main things people actually ask the ai for.

**Comment by Metekillot** [comment]
> -Ideally, we might take door opening away from AIs - Why? Its one of the main things people actually ask the ai for. Tells me you don't understand the appeal of playing ai without telling me. Its not just about job content, Every Ai open is something the ai can remember in their book of grudges if they choose. Its a core part of the appeal of ai.

AI open is a divisive core component and I misspoke in the original post because I was pushing the PR at 2 am:

AI should have the option to not be the only thing that can pass butter on the station. If people want to misuse the laws to make it pass butter that's fine. Letting heads of staff add a law to tell the AI to not pass butter without their approval is a mistake at the moment, because intermittent access is needed in many parts of the gameplay loop. Giving all the heads of staff a butter knife they can use at their leisure is a good compromise and feature add. Their current dinnerware is more like a spoon. Nobody likes to spread butter with a spoon.

**Comment by TheVekter** [comment]
Bisar you annoy me sometimes but this PR is fucking gold and I love it. I actually use the door remote a lot as QM to allow people into the vault for syncies so I'm full steam 100% okay with this change.

**Comment by Us3rn4me228** [comment]
Oh no, they turning Heads into FNAF security guards...

**Comment by SomeoneYouProbablyKnow** [comment]
Got a question? Will this make using the door remote to quickly lockdown doors more difficult?
Having my remote saved me from many an angry mob and if instead of switching from "open" to "bolt" using it opened up a menu for some door request thingy i'd be torn apart. As long as that's still easy to do, this is fire

**Comment by deathrobotpunch** [comment]
I just had an amazing idea, allow door remotes to give the janitor access to departments as well. it’s a pain to use the keycard auth every single time.

**Comment by deathrobotpunch** [comment]
DAMN YOU GITHUB STALE BOT!
